homework/advanced-calculator/advancedCalculator.cpp

```diff
@@ -0,0 +1,139 @@
+                return ErrorCode::DivideBy0;
+            }
+            *out = commands.find('/')->second(lhs, rhs);
+            return ErrorCode::OK;
+            break;
+        case '%':
+            if (!isInteger(lhs) || !isInteger(rhs)){
+                return ErrorCode::ModuleOfNonIntegerValue;
+            }
+            *out = commands.find('%')->second(lhs, rhs);
+            return ErrorCode::OK;
+            break;
+        case '!':
+            *out = commands.find('!')->second(lhs, rhs);
+            return ErrorCode::OK;
+            break;
+        case '^':
+            *out = commands.find('^')->second(lhs, rhs);
+            return ErrorCode::OK;
+            break;
+        case '$':
+            if (lhs <= 0 || rhs == 0) {
+                return ErrorCode::SqrtOfNegativeNumber;
+            }
+            *out = commands.find('$')->second(lhs, rhs);
+            std::cout << *out;
+            return ErrorCode::OK;
+            break;
+        default:
+            return ErrorCode::Undefined;
+            break;
+    }
+}
+
+bool isAllowedChar (char c) {
+    return std::find_if(allowedChars.begin(), allowedChars.end(), [&c](char allowed){ return c == allowed; }) != allowedChars.end();
+}
+
+bool validateCharacters (const std::string & input) {
+    return std::any_of(input.begin(),
+                       input.end(), 
+                       [](char c){ return !isdigit(c) && !isAllowedChar(c) && c != ' ' && c != '.' && c != ','; }); 
+}
+
+bool validateFormat (std::string &input) {
+
+        if (std::count(begin(input), end(input), ',') > 0) {
+            return false;
+        }
+        if (std::count (begin(input), end(input), '-') > 3) {
+            return false;
+        }
+        if (std::count_if(input.begin() + 1, input.end(), [](char c){ return checkFormatSymbols(c) ; }) > 1) {
+            return false;
+        }
+        if (*begin(input) == '+') {
+            return false;
+        }
+        if (std::count (begin(input), end(input), '.') > 2) {
+            return false;
+        }
+        if (findCommand(input, allowedChars) == '!') {
+            auto it = std::find(input.begin(), input.end(), '!');
+            if (it + 1 != input.end()) {
+                return false;
+            }
+        }
+        return true;
+
+}
+
+char findCommand (std::string &input, std::string &allowedChars) {
+    char command;
+    auto it = std::find_first_of(input.begin() + 1,
+                                 input.end(),
+                                 allowedChars.begin(),
+                                 allowedChars.end());
+    return command = *it;
+}
+
+double parseLhs (std::string &input, std::string::size_type &sz) {
+    return std::stod(input, &sz);
+}
+
+double parseRhs (std::string &input, std::string::size_type &sz, char &symbol) {
+    double rhs;
+    if (symbol == '!') {
+        return rhs = 0;
+    }
+    return rhs = std::stod(input.substr(sz+1));
```

Comment:
I saw the sam eimplementation in othe group :)

this is wrong 
if (symbol) {
   return 0;
}
return std::stod(....);